fairlearn/metrics/_metric_frame.py

```diff
@@ -117,24 +126,127 @@ def __init__(self,
                  y_pred, *,
                  sensitive_features,
                  control_features: Optional = None,
-                 sample_params: Optional[Union[Dict[str, Any], Dict[str, Dict[str, Any]]]] = None):
+                 sample_params: Optional[Union[Dict[str, Any], Dict[str, Dict[str, Any]]]] = None,
+                 streaming=False):
         """Read a placeholder comment."""
+        if len(y_true) == 0 and not streaming:
+            raise ValueError(_BAD_INIT_ERROR)
+
```

Comment:
Folks, I'd like to hit a pause and open a quick discussion about what this should look like internally. My suggestion is that we should create an intermediate `DataFrame` representation. Specifically:

**1. We create an internal function of the following form:**

```python
def _pandify(y_true, y_pred, sensitive_features, control_features, sample_params) -> pd.DataFrame:
    
    # My sample code below assumes that y_pred and y_true are 1D array-like,
    # sensitive_features and control_features are dicts of 1D array-like,
    # sample_params are dicts of dicts of 1D array-like.
    # 
    # The actual code would first need to do validation and conversion
    # including the generation of names of control features and sensitive features.
    
    shared_inputs_df = pd.concat(
        {'dummy': pd.DataFrame({'y_true': y_true, 'y_pred': y_pred})}, axis=1)
    sensitive_features_df = pd.concat(
        {'dummy': pd.DataFrame(sensitive_features)}, axis=1)     
    control_features_df = pd.concat(
        {'dummy': pd.DataFrame(control_features)}, axis=1)           
    sample_params_df = pd.concat(
        {key: pd.DataFrame(sample_params[key]) for key in sample_params.keys()}, axis=1)

    combined_df = pd.concat({'shared_inputs': shared_inputs_df,
        'sensitive_features': sensitive_features_df,
        'control_features': control_features_df, 
        'sample_params': sample_params_df}, axis=1)
    return combined_df
```

**2. For non-streaming implementation:**
We call `_pandify` in `__init__` and then calculate `self._overall` and `self._by_group`.

**3. For streaming implementation:**
We call `_pandify` in `add_data`. If this is not the first call, we check that the columns of the new data frame agree with the columns of the old data frame and if they do, we call `pd.concat` on them.

What do you think? (The reason for my micromanaging this is that there are a few extensions of `MetricFrame` in the pipeline and I think they'll be much smoother if we have the intermediate `DataFrame` representation.)

Reply:
There would be some updating in other internal processing functions--we would no longer be passing multiple arguments, e.g., we would have
```python
def _compute_overall(self, func_dict, combined_df):
```

Reply:
That is a great idea! It would simplify a lot of things :) I'll work on it this week.

Reply:
I can see that the internal representation becomes more obscure, but I don't understand the column consistency issue. If we have the current data in `df` and new data in `df_new` we can just check consistency by calling `df.columns.symmetric_difference(df_new.columns).empty` and we can just update with `df = df.append(df_new)`.

Reply:
These operations you're mentioning are quite expensive. And in terms of API design, it's much easier to understand what the input to a method is, if it gets things it needs in separate arguments compared to accepting a single dataframe which the user of the method would need to know how to prepare.

Reply:
Is this issue still outstanding? It does seem quite major based on @adrinjalali 's comments.

Reply:
If I'm not mistaken, we are NOT going down the pandas way, so I'd say we're good?